docs/DesignDocs/PluggableAllocators.md

```diff
@@ -0,0 +1,142 @@
+
+- The design must allow allocators to perform thread-specific optimizations since multi-threaded applications benefit a lot from fine-tuned allocators.
+
+- Allocators must rely only on published interfaces and datastructures.
+
+- The design must allow for Open Enclave core (`liboecore`) to own a portion of the heap and have the allocator manage rest of the heap. </br>
+  This allows the SDK to manage a bit ofmemory for efficiently dispatching ecalls/ocalls, logging and other core features, better than what a general purpose
+  allocator would be capable of. Additionally, it also allows a clear memory demarcation between `core memory` and `application memory`.
+
+- Performance and safety. Plugin approaches often involve registering `hook` functions that will be called by the SDK via an indirect calls.
+  It is desirable to avoid indirect calls since they are not performant. Additionally, indirect calls are subject to speculative execution side-channels.
+
+- Backwards-compatibility. No existing clients must require any change to get the default allocator.
+
+
+Design
+---
+
+A custom allocator must provide a set of replacement functions. They are declared in the published header `openenclave/enclave/allocator.h`.
+The SDK will provide default implementations of these functions that will result in the use of the default allocator.
+
+If no replacement functions are provided, then the default allocator is used. This retains the current behavior in which all enclaves get the default allocator without any additional linker options.
+
+The replacement functions can be provided
+- via object files. This happens when the allocator is not a separate static library.
+- via a separate allocator static library. In this case, the allocator library must be specified before `oelibc` in the linker command line. </br>
+  ELF  `weak symbols` (see [Function Attributes](https://gcc.gnu.org/onlinedocs/gcc-3.2/gcc/Function-Attributes.html)) are designed to allow user code to override functions defined in libraries.
+  Using `weak symbols` for default implementation of these function would cause the linker to look for custom replacement functions irrespective of the order.
+
+
+Replacement functions are not required to be reentrant.
+
+Undefined Behavior
+---
+
+- The replacement functions should not do much other than managing memory. Making ocalls, logging, printf etc may use dynamic memory and hence result in reentranct calls to the allocator.
+- All the replacement functions must be provided, or none. A partial set will result in both the default allocator and the custom allocator managing the heap.
+
+
+Replacement Functions
+---
+
+Published in `openenclave/enclave/allocator.h`
+
+- `void oe_allocator_init(void* heap_start_address, void* heap_end_address)` </br>
+  This function will be called by `oecore` at an appropriate time during enclave initialization. The total heap memory available to the enclave is specified in its configuration file
```

Comment:
The part about "in its configuration file" is not true, when OE_SET_ENCLAVE_SGX or OE_SET_ENCLAVE_OPTEE are used instead.  (See also issue #2714)